Add plumbing to import WebGPU::Queue and WebGPU::Buffer into Swift as reference types

nmahendru opened the pull request:
#### d95821ede59f3e9ec15f2a11437f6b748cc9bcf7
<pre>
Add plumbing to import WebGPU::Queue and WebGPU::Buffer into Swift as reference types
<a href="https://bugs.webkit.org/show_bug.cgi?id=280409">https://bugs.webkit.org/show_bug.cgi?id=280409</a>
<a href="https://rdar.apple.com/136698682">rdar://136698682</a>

Reviewed by Mike Wyrzykowski.

This change can be considered as a template to importing WebKit Reference types into Swift.
Using SWIFT_SHARED_REFERENCE with a type ensures that:
1. Swift holds a strong reference to the type passed in guaranteeing lifetime.
2. Swift will manage releasing the reference once it&apos;s done with that object.

Before this Change, it was impossible to import any RefCounted types as Swift by default will only
import Copy types and move types in certain cases. But for WebGPU work, this is really useful.

There are some extern C related changes done to keep header includes outside of extern C which is
needed for Swift Cpp Interop.

Other Changes:
1. Now Queue.h and Buffer.h are Project headers.
2. A new preprocessor define __WEBGPU__ has been added to guard these headers so they are only visible in
   WebGPU.framework. The headers needed to be marked Project to keep Swift tapi generation happy.
3. This all is still disabled by default.

Test:
Build: make release BUILD_WEBKIT_OPTIONS=&quot;--webGPUSwift&quot;
and then run all the tests under http/tests/webgpu/webgpu/api/operation/queue http/tests/webgpu/webgpu/api/validation/queue.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/RefCounted.h:
(retainRefCounted):
(releaseRefCounted):
* Source/WTF/wtf/RetainReleaseSwift.h: Added.
* Source/WTF/wtf/ThreadSafeRefCounted.h:
(retainThreadSafeRefCounted):
(releaseThreadSafeRefCounted):
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
(retainBuffer):
(releaseBuffer):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Queue.h:
(retainQueue):
(releaseQueue):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
* Source/WebGPU/WebGPU/Queue.swift: Added.
(writeBuffer(_:buffer:offset:data:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/bmalloc/libpas/src/libpas/pas_utils.h:

Canonical link: <a href="https://commits.webkit.org/284516@main">https://commits.webkit.org/284516@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c095d29a31d53962e55a113d710fca04cd78ef5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69689 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22440 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/20847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56890 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20698 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/73774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/20847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72755 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44754 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/60108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41420 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/17547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19224 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/62805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63355 "Build is in progress. Recent messages:") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17892 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75487 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68935 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13671 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/17156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13711 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60191 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62975 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10993 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/4583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90717 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10630 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44893 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16103 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45967 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/47231 "Build was cancelled. Recent messages:OS: Sonoma (14.5), Xcode: 15.4") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45708 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.h:
```diff
@@ -148,6 +152,23 @@ class Buffer : public WGPUBufferImpl, public RefCounted<Buffer>, public CanMakeW
 #endif
+#if ENABLE(WEBGPU_SWIFT)
+} SWIFT_SHARED_REFERENCE(retainBuffer, releaseBuffer);
+#else
```
I believe SWIFT_SHARED_REFERENCE is already a macro that compiles to nothing if you're not building with Swift-C++ interop enabled. If it's not, let's make a WebKit macro.

The we can just say "MACRO(retainBuffer, releaseBuffer)", without the ENABLE guard.

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.h:
```diff
@@ -148,6 +152,23 @@ class Buffer : public WGPUBufferImpl, public RefCounted<Buffer>, public CanMakeW
 
 } // namespace WebGPU
+
+#if ENABLE(WEBGPU_SWIFT)
```
No need for macro here.

geoffreygaren commented on Source/WebGPU/WebGPU/Queue.h:
```diff
@@ -118,6 +121,20 @@ class Queue : public WGPUQueueImpl, public ThreadSafeRefCounted<Queue> {
+#endif
 } // namespace WebGPU
+
+#if ENABLE(WEBGPU_SWIFT)
```
No need for macro here.

geoffreygaren commented on Source/WebGPU/WebGPU/Queue.swift:
```diff
@@ -0,0 +1,29 @@
+ */
+
+
+public func writeBufferSwift(queue: WebGPU.Queue, buffer: WebGPU.Buffer, offset: UInt64, data: SpanUInt8) {
```
I'd like to avoid putting "Swift" in the names of all our Swift functions. We don't put "ObjC" in the names of all our ObjC functions, or "CPP" in the names of all our C++ functions.

Is overload resolution not sufficient to say just `writeBuffer`?

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.h:
```diff
@@ -148,6 +152,23 @@ class Buffer : public WGPUBufferImpl, public RefCounted<Buffer>, public CanMakeW
 } // namespace WebGPU
+
+#if ENABLE(WEBGPU_SWIFT)
+inline void retainBuffer(WebGPU::Buffer* obj)
```
FYI: Swift-C++ has a bug where the very first refcount on a WebKit object will be wrong. So, for the time being, you need to avoid allocating refcounted objects from Swift. I believe the m_adoptionIsRequired ASSERT inside ref() is already sufficient to catch this bug in debug builds. So we don't need to do anything special yet.

nmahendru replied:
thanks for mentioning that!

mwyrzykowski replied:
Maybe I don't have all the context, but do we need these `SWIFT_SHARED_REFERENCE` macros if we shouldn't be allocating refcounted objects from Swift?

I think I am missing some context as I am not sure where they are used.

geoffreygaren reviewed (changes requested):
Looks good but I have some suggestions for improvements.

mwyrzykowski commented on Source/WebGPU/WebGPU/Instance.h:
```diff
@@ -34,6 +34,8 @@
 #import <wtf/ThreadSafeRefCounted.h>
 #import <wtf/WeakPtr.h>
+#include "WebGPU.h"
+#include "WebGPUExt.h"
```
not strictly necessary, but generally WebGPU.h / WebGPUExt.h is imported like:

```
#import <WebGPU/WebGPU.h>
#import <WebGPU/WebGPUExt.h> 
```
but perhaps with continued changes to these headers, we should stop doing that.

Probably changing `include` -> `import` would be reasonable

mwyrzykowski commented on Source/bmalloc/libpas/src/libpas/pas_utils.h:
```diff
@@ -53,8 +53,6 @@ PAS_IGNORE_CLANG_WARNINGS_BEGIN("qualifier-requires-header")
 #define PAS_BEGIN_EXTERN_C __PAS_BEGIN_EXTERN_C
 #define PAS_END_EXTERN_C __PAS_END_EXTERN_C
 
-PAS_BEGIN_EXTERN_C;
```
why is this change needed in `bmalloc`?

nmahendru replied:
IIRC some transitive include from pas_utils_additions.h has a #include <> inside extern "C" which the swift compiler does not like.
more context:
https://forums.swift.org/t/import-of-c-module-appears-within-extern-c-language-linkage-specification/65606

geoffreygaren commented on Source/WebGPU/WebGPU/Buffer.h:
```diff
@@ -148,6 +151,17 @@ class Buffer : public WGPUBufferImpl, public RefCounted<Buffer>, public CanMakeW
+
+inline void retainBuffer(WebGPU::Buffer* obj)
+{
+    WTF::RetainReleaseHelperRefCounted::retain(obj);
```
What prevents us from just calling obj->ref() here?

I believe we can remove the RetainReleaseHelper classes. If we can static_assert that T is derived from RefCounted, that means the compiler can see that T derives from RefCounted at compile time, so RefCounted::ref() should be in scope.

(Apologies if I'm repeating this comment. I feel like I said it before, but I don't see it anywhere.)

nmahendru replied:
This is basically a placeholder for if we want to do something custom for retain release. Like logging or some workarounds.
Right now we have 2 classes getting imported but that number will go up significantly.
All of the quirks can just be in the helper class and not at callsites

mwyrzykowski replied:
Do we need these manual definitions in each header file?

Couldn't it be directly integrated into RefCounted.h for instance?
```
template<typename T>
static void retainRefCounted(T* obj)
{
...
}
template<typename T>
static void releaseRefCounted(T* obj)
{
...
}
```

that way we don't need manual definitions in each header file?

mwyrzykowski reviewed (commented):
If we merged the separate Queue and Buffer retain / release functions into a single static function, it seems like we could also take @geoffreygaren's feedback and simply have them call ref() and deref() until a more complex implementation is needed

mwyrzykowski reviewed (approved):
Thank you!

nmahendru committed:
Add plumbing to import WebGPU::Queue and WebGPU::Buffer into Swift as reference types
https://bugs.webkit.org/show_bug.cgi?id=280409
rdar://136698682

Reviewed by Mike Wyrzykowski.

This change can be considered as a template to importing WebKit Reference types into Swift.
Using SWIFT_SHARED_REFERENCE with a type ensures that:
1. Swift holds a strong reference to the type passed in guaranteeing lifetime.
2. Swift will manage releasing the reference once it's done with that object.

Before this Change, it was impossible to import any RefCounted types as Swift by default will only
import Copy types and move types in certain cases. But for WebGPU work, this is really useful.

There are some extern C related changes done to keep header includes outside of extern C which is
needed for Swift Cpp Interop.

Other Changes:
1. Now Queue.h and Buffer.h are Project headers.
2. A new preprocessor define __WEBGPU__ has been added to guard these headers so they are only visible in
   WebGPU.framework. The headers needed to be marked Project to keep Swift tapi generation happy.
3. This all is still disabled by default.

Test:
Build: make release BUILD_WEBKIT_OPTIONS="--webGPUSwift"
and then run all the tests under http/tests/webgpu/webgpu/api/operation/queue http/tests/webgpu/webgpu/api/validation/queue.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/RefCounted.h:
(retainRefCounted):
(releaseRefCounted):
* Source/WTF/wtf/RetainReleaseSwift.h: Added.
* Source/WTF/wtf/ThreadSafeRefCounted.h:
(retainThreadSafeRefCounted):
(releaseThreadSafeRefCounted):
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
(retainBuffer):
(releaseBuffer):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Queue.h:
(retainQueue):
(releaseQueue):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::writeBuffer):
* Source/WebGPU/WebGPU/Queue.swift: Added.
(writeBuffer(_:buffer:offset:data:)):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/bmalloc/libpas/src/libpas/pas_utils.h:

Canonical link: https://commits.webkit.org/284516@main
